### DISCUSSION TITLE
Parsing of negative zero

### DISCUSSION THREAD

**Discussion by 6502** [nlohmann/json#2854] [Q&A] [answered]
I noticed that parsing `-0` results in `0`, is this intentional or just happened? I saw nothing in the documentation or in the issue list about it...

**Comment by nlohmann** [comment, marked as answer]
Number handling is described in https://json.nlohmann.me/features/types/number_handling/ - apparently, `-0` is allowed, but is treated like `0` by this library. What would you expect instead?

**Reply by 6502** [reply]
-0.0 parses as negative zero (yes, IEEE754  has negative zero that is distinct from zero - even if they must compare equal with ==). I'm not opinionated on this but found this little difference between your parser and mine (and JSON formally doesn't know what an integer is)

**Reply by 6502** [reply]
By the way checking other libraries seems this is a very common behavior (for example both rapidjson and the json parser of the standard library of python do the same). I would say that parsing -0 as zero (and instead -0.0 as negative zero) is incorrect (in other words I see a similar case for e.g. 9999999999999999999999999... if the number cannot be represented correctly as an int you parse it as a double; -0 cannot be represented as an int, so I'd say should be represented in a double). JSON is however about interoperability first, so if most other libraries are parsing -0 as 0 (throwing away the negative information) then may be it is a better idea even if IMO more confused and questionable in abstract given that JSON knows nothing about integer/float differences. If you like I can propose a documentation integration for at least describing this behavior.

**Reply by nlohmann** [reply]
I would add the following paragraph to the documentation

* * *

### Zeros

The JSON number grammar allows for different ways to express zero, and this library will store zeros differently:

| Literal | Stored value and type | Serialization |
| ------- | --------------------- | ------------- |
| `0`     | `std::uint64_t(0)` | `0`          |
| `-0`    | `std::int64_t(0)` | `0`           |
| `0.0`   |  `double(0.0)`     | `0.0`         |
| `-0.0`  | `double(-0.0)`    | `-0.0`        |
| `0E0`   | `double(0.0)`     | `0.0`         |
| `-0E0`  | `double(-0.0)`    | `-0.0`        |

That is, `-0` is stored as a signed integer, but the serialization does not reproduce the `-`.

**Reply by 6502** [reply]
That clarifies and indeed even allows to distinguish `0` and `-0` in the improbable case someone actually cares. In other words the information is not actually lost as I originally thought. :+1:
